src/CoreWCF.Primitives/src/CoreWCF/Security/SpnegoTokenAuthenticator.cs

```diff
@@ -0,0 +1,185 @@
+          //  if (this.credentialsHandle == null)
+          //  {
+          //      this.credentialsHandle = SecurityUtils.GetCredentialsHandle("Negotiate", this.serverCredential, true);
+          //  }
+          if(this.negotiateHandler ==null)
+                this.negotiateHandler = (NegotiateInternal.NegotiateInternalState)new NegotiateInternal.NegotiateInternalStateFactory().CreateInstance();
+
+            return Task.CompletedTask;
+        }
+
+        public override Task CloseAsync(CancellationToken token)
+        {
+            base.CloseAsync(token);
+            FreeCredentialsHandle();
+            return Task.CompletedTask;
+        }
+
+        public override void OnAbort()
+        {
+            try
+            {
+                base.OnAbort();
+            }
+            finally
+            {
+                FreeCredentialsHandle();
+            } 
+        }
+
+        void FreeCredentialsHandle()
+        {
+            //if (this.credentialsHandle != null)
+            //{
+            //    this.credentialsHandle.Close();
+            //    this.credentialsHandle = null;
+            //}
+            if (this.negotiateHandler != null)
+                this.negotiateHandler.Dispose();
+        }
+
+        protected override SspiNegotiationTokenAuthenticatorState CreateSspiState(byte[] incomingBlob, string incomingValueTypeUri)
+        {
+            ISspiNegotiation windowsNegotiation = new WindowsSspiNegotiation("Negotiate", DefaultServiceBinding, GetNegotiateState());
+            return new SspiNegotiationTokenAuthenticatorState(windowsNegotiation);
+        }
+
+        protected override ReadOnlyCollection<IAuthorizationPolicy> ValidateSspiNegotiation(ISspiNegotiation sspiNegotiation)
+        {
+            WindowsSspiNegotiation windowsNegotiation = (WindowsSspiNegotiation)sspiNegotiation;
+            if (windowsNegotiation.IsValidContext == false)
+            {
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperWarning(new SecurityNegotiationException(SR.Format(SR.InvalidSspiNegotiation)));
+            }
+           // SecurityTraceRecordHelper.TraceServiceSpnego(windowsNegotiation);
+            if (this.IsClientAnonymous)
+            {
+                return EmptyReadOnlyCollection<IAuthorizationPolicy>.Instance;
+            }
+            IIdentity identity = windowsNegotiation.GetIdentity();
+            if (identity != null)
+            {
+                return GetAuthorizationPolicies(identity);
+            }
+            else
+                throw new Exception("Identity can't be determined.");
+            //using (SafeCloseHandle contextToken = windowsNegotiation.GetContextToken())
+            //{
+            //    WindowsIdentity windowsIdentity = new WindowsIdentity(contextToken.DangerousGetHandle(), windowsNegotiation.ProtocolName);
+            //    SecurityUtils.ValidateAnonymityConstraint(windowsIdentity, this.AllowUnauthenticatedCallers);
+
```

Comment:
Not doing the equivalent of this might be a security issue.

Reply:
I am doing below.. just moved to below method.. I will remove the comments not to confuse